lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp

```diff
@@ -264,6 +266,12 @@ void IMDeadCodeElimPass::markBlockExecutable(Block *block) {
     else if (hasUnknownSideEffect(&op))
       markUnknownSideEffectOp(&op);
 
+    // Recursively mark any blocks contained within these operations as
+    // executable.
+    for (auto &region : op.getRegions())
+      for (auto &block : region.getBlocks())
+        markBlockExecutable(&block);
```

Comment:
I think we should move this under `hasUnknownSideEffect` op.  I consider this the conservative codepath for when we don't know what an operation is. Layerblocks fall under this umbrella.

If we ever had an operation like this totally made up example:
```
sv.ifdef 0 {
  sv.assert(clock, 0)
}
```
We would not want to unconditionally visit its regions.

Reply:
I think this seems right. It seems like the better thing would be to make this the default for any unknown operation and not just `hasUnknownSideEffect`. That is different from how the PR is written, though, and means that there should be early exits from all the prior `if`/`else if` clauses.

I.e., the original problem with this pass was that unless this is the fallback _for all unknown operations_ then the pass will incorrectly delete declarations without deleting their uses.

Switched to your suggestion in: https://github.com/llvm/circt/pull/7881/commits/c4d490a51d536bfdd11cf11ab275522de081f32e